storage-node/src/cache/data_store_cache/memdisk/mod.rs

```diff
@@ -225,44 +323,87 @@ impl<R: DataStoreReplacer<MemDiskStoreReplacerKey, MemDiskStoreReplacerValue>> D
                 self.disk_store
                     .write_data(disk_store_key.clone(), Some(bytes_vec), None)
                     .await?;
-                let mut disk_replacer = self.disk_replacer.write().await;
+                let mut disk_replacer = self.disk_replacer.lock().await;
                 if disk_replacer
                     .put(
                         remote_location_evicted,
                         MemDiskStoreReplacerValue::new(disk_store_key.clone(), data_size),
                     )
                     .is_none()
                 {
-                    self.disk_store.clean_data(&disk_store_key).await?;
+                    if let Err(e) = self.disk_store.clean_data(&disk_store_key).await {
+                        warn!(
+                            "Failed to clean data ({}) from disk store: {}",
+                            disk_store_key, e
+                        );
+                    }
+                    warn!(
+                        "Failed to put evicted data ({}) to disk replacer.",
+                        disk_store_key
+                    );
                 }
             }
         }
 
         // 3. If the data is successfully written to memory, directly return.
         if self.mem_store.is_some() && bytes_to_disk.is_none() {
+            let mut status_of_keys = self.status_of_keys.write().await;
+            let ((status, size), notify) = status_of_keys.get_mut(&remote_location).unwrap();
+            *status = Status::Completed;
+            *size = bytes_mem_written;
+            {
+                let mut mem_replacer = self.mem_replacer.as_ref().unwrap().lock().await;
+                mem_replacer.pin(&remote_location, *notify.1.lock().await);
+            }
+            notify.0.notify_waiters();
             return Ok(bytes_mem_written);
```

Comment:
Shall we adopt the same logic to ensure the atomicity here as well?

Reply:
`let mut status_of_keys = self.status_of_keys.write().await;` has already ensure the atomicity I guess lol

Reply:
Also, I want to move `notify.0.notify_waiters();` out of the scope of status_of_keys write lock, so waking threads can immediately access the lock?